site/_posts/2020-05-20-building-protected-api-with-rails-and-jwt.md

```diff
@@ -0,0 +1,254 @@
+
+The first step to changing this API is to write tests. Let's modify the test to provide a JWT and expect `:forbidden` HTTP statuses when the token doesn't meet our expectations.
+
+```ruby
+class MessagesTest < ActionDispatch::IntegrationTest
+  test "can' get messages with no auth" do
+    get "/messages"
+    assert_response :forbidden
+  end
+  test "can get messages with header" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt }
+    assert_response :success
+  end
+  test "expired jwt fails" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt(-1) }
+    assert_response :forbidden
+  end
+  test "can get messages content" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt }
+    res = JSON.parse(@response.body)
+    assert_equal '{"messages"=>["Hello"]}', res.to_s
+  end
+
+  def build_jwt(valid_for_minutes = 5)
+    exp = Time.now.to_i + (valid_for_minutes*60)
+    payload = { "iss": "fusionauth.io",
+                "exp": exp,
+                "aud": "238d4793-70de-4183-9707-48ed8ecd19d9",
+                "sub": "19016b73-3ffa-4b26-80d8-aa9287738677",
+                "name": "Dan Moore",
+                "roles": ["USER"]
+    }
+
+    JWT.encode payload, Rails.configuration.x.oauth.jwt_secret, 'HS256'
+
+  end
+end
+```
+
+What we're doing here is writing a test that specifies the JWT will be in the `Authorization` header. Let's look more closely at the token. 
+
+[JWTs have claims](https://tools.ietf.org/html/rfc7519#section-4), basically information embedded in the JWT. The keys of the JSON payload we build in the `build_jwt` function, such as `iss` and `name`, are claims. Some of these are defined in the JWT RFC. These are 'registered' claims. Others are recorded with the IANA but are not part of the standard; these are 'public' claims. And yet others are added to the payload by the token creator; these are 'private' claims. 
+
+
+```ruby
+# ...
+  def build_jwt(valid_for_minutes = 5)
+    exp = Time.now.to_i + (valid_for_minutes*60)
+    payload = { "iss": "fusionauth.io",
+                "exp": exp,
+                "aud": "238d4793-70de-4183-9707-48ed8ecd19d9",
+                "sub": "19016b73-3ffa-4b26-80d8-aa9287738677",
+                "name": "Dan Moore",
+                "roles": ["USER"]
+    }
+
+    JWT.encode payload, Rails.configuration.x.oauth.jwt_secret, 'HS256'
+# ...
+```
+
+Above, we create registered claims that our API may examine. `exp` indicates when the JWT will expire. `aud` is an identifier of who or what this token is intended for (the "audience"). `sub` is the person or piece of software this token applies to--to quote the RFC: "The claims in a JWT are normally statements about the subject." `iss` is an identifier for the issuer of the JWT. 
```

Comment:
When rendered, the to--to doesn't read well. The double dash doesn't show up so it looks like to-to. Perhaps use parenthesis or something like that?